### PR TITLE
[io-layer] Extract threading code

### PR DESCRIPTION
**Pull request by ludovic-henry**
cc @kumpera @lambdageek

### CODE REVIEW

**Comment by ludovic-henry** [comment]
build

File: mono/io-layer/wait.c [hunk 1]
```diff
@@ -68,7 +69,7 @@ guint32 WaitForSingleObjectEx(gpointer handle, guint32 timeout,
 	guint32 ret, waited;
 	int thr_ret;
 	gboolean apc_pending = FALSE;
-	gpointer current_thread = wapi_get_current_thread_handle ();
+	gpointer current_thread = mono_thread_info_get_handle (mono_thread_info_current ());
 	gint64 now, end;
 	
 	if (current_thread == NULL) {
```
Comment (lambdageek) [inline mono/io-layer/wait.c]: isn't this condition always false now?  you assert in mono_thread_info_get_handle that info->handle isn't null.

File: mono/mini/mini-exceptions.c [hunk 2]
```diff
@@ -2508,9 +2505,9 @@ mono_print_thread_dump_internal (void *sigctx, MonoContext *start_ctx)
 		g_string_append (text, "\n\"<unnamed thread>\"");
 
 #ifndef HOST_WIN32
```
Comment (lambdageek) [inline mono/mini/mini-exceptions.c]: now that you have a stub mono_threads_platform_describe on win32, does it make sense to remove this #ifndef ?

File: mono/io-layer/wait.c [hunk 3]
```diff
@@ -244,21 +234,7 @@ guint32 SignalObjectAndWait(gpointer signal_handle, gpointer wait,
...
 	gboolean apc_pending = FALSE;
-	gpointer current_thread = mono_thread_info_get_handle (mono_thread_info_current ());
 	gint64 now, end;
-	
-	if (current_thread == NULL) {
-		SetLastError (ERROR_INVALID_HANDLE);
-		return(WAIT_FAILED);
-	}
-
-	if (signal_handle == _WAPI_THREAD_CURRENT) {
```
Comment (lambdageek) [inline mono/io-layer/wait.c]: Why can't the signal handle be _WAPI_THREAD_CURRENT?

File: mono/io-layer/wait.c [hunk 4]
```diff
@@ -462,10 +432,6 @@ guint32 WaitForMultipleObjectsEx(guint32 numobjects, gpointer *handles,
 
 	/* Check for duplicates */
 	for (i = 0; i < numobjects; i++) {
-		if (handles[i] == _WAPI_THREAD_CURRENT) {
```
Comment (lambdageek) [inline mono/io-layer/wait.c]: Why can't handles[i] be _WAPI_THREAD_CURRENT?

File: mono/io-layer/wait.c (hunk 1 above)
Comment (ludovic-henry) [inline mono/io-layer/wait.c]: that's removed in a later commit, but you are right, I will delete it in this commit.

**Comment by lambdageek** [comment]
You removed a bunch of places where the code checks for `_WAPI_THREAD_CURRENT`, but as far as I know, the win32 `GetCurrentThread()` will return that value.  Is it the case that on non-win32 we'll just never return that constant, so the functions don't need to handle it; and on win32 it will use the regular win32 api which will presumably do the right thing with that constant?

File: mono/io-layer/wait.c (hunk 3 above)
Comment (ludovic-henry) [inline mono/io-layer/wait.c]: We do not create handle with value _WAPI_THREAD_CURRENT any more. So no handle can have this value.

File: mono/io-layer/wait.c (hunk 4 above)
Comment (ludovic-henry) [inline mono/io-layer/wait.c]: Same as above, we never assign the handle value _WAPI_THREAD_CURRENT.

**Comment by ludovic-henry** [comment]
`GetCurrentThread` would only valid on win32 platform after this change, and the place where it could return `_WAPI_THREAD_CURRENT`, would be in the io-layer which is not used on win32. 

Moreover, when creating/attaching a thread, we now allocate a real handle (and not a "fake" one like `_WAPI_THREAD_CURRENT`) as you can see in `mono_threads_platform_register`. Also on win32, we explicitely call `DuplicateHandle` on the value returned by `GetCurrentThread` to get a "real" handle.

**Comment by lambdageek** [comment]
There are Windows compilation failures.  Otherwise LGTM.

File: mono/utils/mono-threads.h [hunk 5]
```diff
@@ -289,6 +297,14 @@ typedef struct {
 	guint32 stack_size;		
 } MonoThreadParm;
 
+typedef enum {
+	THREAD_PRIORITY_LOWEST       = 0,
```
Comment (lambdageek) [inline mono/utils/mono-threads.h]: These constants are already defined on Win32

File: mono/metadata/domain.c [hunk 6]
```diff
@@ -526,7 +526,6 @@ mono_init_internal (const char *filename, const char *exe_filename, const char *
 #endif
 
 #ifndef HOST_WIN32
-	mono_w32handle_init ();
 	wapi_init ();
 #endif
```
Comment (vargaz) [inline mono/metadata/domain.c]: Why is this needed ? Traditionally, we tried to move away from lazy initialization since it complicates things.

Comment (ludovic-henry) [inline mono/metadata/domain.c]: this is needed because the profiler might try to create a new handle before having called `mono_init_internal`. I am going to revert that change anyway.
